Modem: Add pipelink module

bjarki-andreasen opened the pull request:
The pipelink module allows modem drivers to expose modem pipes globally, advertising when pipes become available.
This will allow us to split the cellular modem driver into multiple bespoke modules, each handling a specific feature, like PPP, socket offloading, AT shell, custom application logic etc. See #73722 for an overview of its place in the overall modem device driver design.

Modem drivers will be tasked with setting up the modem, creating as many pipes as the application/subsystems need, then advertising that when the pipes are ready to be opened.

To demonstrate the use of pipelinks, the modem_cellular driver has been updated to create set up two additional DLCI channels for use by the application, named `user_pipe_0` and `user_pipe_1` (the names can be anything)

The driver advertises that the two additional pipes are ready to be used once CMUX has been connected. A shell command for sending AT commands through one of these user pipes has been added to demonstrate the feature.

A small demo (you can also test out the pull request locally of course :))
Turn on the modem using the pm shell command
```
uart:~$ pm resume modem
...
[00:07:31.885,559] <inf> modem_at_shell: pipe connected
[00:07:31.885,620] <inf> modem_at_shell: opening pipe
[00:07:31.907,836] <inf> modem_at_shell: pipe opened
[00:07:31.907,928] <inf> modem_at_shell: chat attached

```
The at shell is now ready, simply type `modem at <command>`, in this example "at":
```
uart:~$ modem at at
OK
uart:~$
```
We can send something a bit more interresting returning multiple lines, like "ATI"
```
uart:~$ modem at ATI
Quectel
BG95-M3
Revision: BG95M3LAR02A03
OK
uart:~$ 
```
The at shell is simply creating a dynamic chat script using the modem_chat module, and running it, this is the log output from running "ATI"
```
[00:10:15.743,957] <dbg> modem_chat: modem_chat_script_start: running script: at_shell_script
[00:10:15.743,988] <dbg> modem_chat: modem_chat_script_next: at_shell_script: step: 0
[00:10:15.744,049] <dbg> modem_chat: modem_chat_script_next: sending: ATI
[00:10:15.766,571] <dbg> modem_chat: modem_chat_log_received_command:  Quectel
[00:10:15.768,951] <dbg> modem_chat: modem_chat_log_received_command:  BG95-M3
[00:10:15.771,911] <dbg> modem_chat: modem_chat_log_received_command:  Revision: BG95M3LAR02A03
[00:10:15.776,092] <dbg> modem_chat: modem_chat_log_received_command: OK
[00:10:15.778,900] <dbg> modem_chat: modem_chat_script_stop: at_shell_script: complete
```
Any other custom logic, sending custom commands, can be implemented properly just like the AT shell has been.

Now, replace the at shell with a reusable PPP implementation, an AT command interface for the cellular.h API, NMEA0183 parser for a potential GNSS, any and all application specific logic. The potential to make something truly modular is there :)

bjarki-andreasen committed:
modem: add modem_pipelink module

Add modem pipelink module which exposes modem pipes globally.
The pipelink module implements a callback to inform when a
pipe becomes available to use by whichever modem is attached
to it. This could be a shell, or a network interface.

The module aims to allow modem drivers to be split into modules,
and allowing applications to implement their own custom logic
without altering the modem drivers.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
modem: drivers: add user pipes to cellular modem driver

Add two DLCI channels to the cellular modem driver, and expose
them using the modem pipelink module.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

tomi-font commented on include/zephyr/modem/pipelink.h:
```diff
@@ -0,0 +1,150 @@
+ * @param link Pipelink instance
+ * @retval Pointer to pipe if pipelink has been initialized
+ * @retval NULL if pipelink has not been initialized
+ * @note Pipelink is initialized on first connected event
```
What do you mean with this?

bjarki-trackunit replied:
Updated the note to `Pipelink is guaranteed initialized after first connected event`. The point being that if you wait for the first connected event, `modem_pipelink_get_pipe()` will not return NULL, so you don't have to check the return every time :)

tomi-font replied:
Okay. Though the first connected event doesn't do have anything to do with `link->pipe`, as it's set already and only in `modem_pipelink_init()`?

bjarki-trackunit replied:
Correct :) Maybe the note is implied? Just thought it may be helpful

tomi-font replied:
I rather find the note confusing in that case, talking about connected events while in practice they have nothing to do?

bjarki-trackunit replied:
I have removed the note :)

tomi-font commented on subsys/modem/modem_pipelink.c:
```diff
@@ -0,0 +1,74 @@
+{
+	K_SPINLOCK(&link->spinlock) {
+		link->connected = false;
+		try_callback(link, MODEM_PIPELINK_EVENT_DISCONNECTED);
```
Only call the callback if `connected` is not already equal to the event being sent to prevent duplicate calls?

bjarki-trackunit replied:
Updated to prevent duplicate calls

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -88,11 +89,17 @@ struct modem_cellular_data {
 	/* DLCI 2 is only used for chat scripts. */
 	uint8_t dlci2_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
+	uint8_t dlci3_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
+	uint8_t dlci4_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
```
We should likely do something about the buffer sizes... At least this can't be used like that when extending the number of DLCIs and their use case.

bjarki-trackunit replied:
A new config named MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES has been added for the generic user pipes

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -1560,6 +1581,32 @@ static int modem_cellular_init(const struct device *dev)
+		data->dlci4_pipe = modem_cmux_dlci_init(&data->cmux, &data->dlci4,
+							&dlci4_config);
+
+		modem_pipelink_init(config->shell_pipelink, data->dlci4_pipe);
```
```suggestion
		modem_pipelink_init(config->gnss_pipelink, data->dlci4_pipe);
```
Also, the DLCI initialization could be put into a helper function.

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -2275,6 +2355,8 @@ MODEM_CHAT_SCRIPT_DEFINE(sqn_gm02s_periodic_chat_script,
 		.dial_chat_script = &swir_hl7800_dial_chat_script,                                 \
 		.periodic_chat_script = &swir_hl7800_periodic_chat_script,                         \
+		.shell_pipelink = MODEM_PIPELINK_DT_GET(DT_DRV_INST(inst), shell),                 \
+		.gnss_pipelink = MODEM_PIPELINK_DT_GET(DT_DRV_INST(inst), gnss),                   \
```
Not introduced by this PR, but I'm thinking it could be nice to put back the parts common to all supported modems in a single place instead of having to repeat them for every modem.

bjarki-trackunit replied:
Agree, my plan is to refactor this driver to be split into smaller modules, connected by modem pipelinks, which should remove a bunch of duplicate code from this driver :) but, in a future PR

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -88,11 +89,17 @@ struct modem_cellular_data {
 	/* DLCI 2 is only used for chat scripts. */
 	uint8_t dlci2_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
+	uint8_t dlci3_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
+	uint8_t dlci4_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
```
Actually, if the goal is only to expose DLCIs for whatever else to use, could we make it so that the DLCIs would first exist without receive buffer, and the actual user of every DLCI would attach a buffer to it? This would prevent coupling of the receive buffers to the code that manages the CMUX link.

bjarki-trackunit replied:
The issue is the layering of pipes, a user of a pipe, like a chat module, does not know if it is connected to a UART, or a DLCI channel, very much by design :) UART requires a transmit and receive buffer, DLCI only requires a receive buffer. It would be a layering violation for the user of a pipe to need to know how it is implemented, which is required if the user has to pass a buffer/buffers of appropriate sizes :)

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -2350,6 +2442,9 @@ MODEM_CHAT_SCRIPT_DEFINE(sqn_gm02s_periodic_chat_script,
 	};                                                                                         \
                                                                                                    \
+	MODEM_PIPELINK_DT_DEFINE(DT_DRV_INST(inst), shell);                                        \
+	MODEM_PIPELINK_DT_DEFINE(DT_DRV_INST(inst), gnss);                                         \
```
It would be nice to have at least a reference of how things are meant to be defined in the devicetree.

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+
+	pipe = modem_pipelink_get_pipe(shell_pipelink);
+	if (pipe == NULL) {
+		printk("Nearly exploded, why is pipe NULL?");
```
Hehehe. :smiley:

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+	SHELL_SUBCMD_SET_END
+);
+
+SHELL_CMD_REGISTER(at_shell, &sub_at_shell_cmds, "Modem AT shell commands", NULL);
```
Is it possible to make it so that, instead of the command being `at_shell at AT+MYCOMMAND`, it is directly `at[_shell] AT+MYCOMMAND`?

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+		printk("AT shell connected\n");
+		break;
+
+	case MODEM_PIPELINK_EVENT_DISCONNECTED:
```
What happens to the pipe when this is received, who closes it?

bjarki-trackunit replied:
The pipe is forcefully closed by the provider of the pipe. In the case of CMUX, the disconnected event would precede the CMUX being disconnected.

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -1560,6 +1581,32 @@ static int modem_cellular_init(const struct device *dev)
+	{
+		const struct modem_cmux_dlci_config dlci3_config = {
+			.dlci_address = 3,
+			.receive_buf = data->dlci2_receive_buf,
```
```suggestion
			.receive_buf = data->dlci3_receive_buf,
```
Having those in an array that is looped over could be more handy (and prevent such errors).

bjarki-trackunit replied:
Agreed, I'm working on a proper sample that creates a bunch of DLCI channels for pipelinks, I'm doing the following to prevent these errors:
```
struct bg9x_pipe {
	struct modem_cmux_dlci dlci;
	uint8_t dlci_address;
	uint8_t *dlci_receive_buf;
	uint16_t dlci_receive_buf_size;
	struct modem_pipe *pipe;
	struct modem_pipelink *pipelink;
};

static void bg9x_init_pipe(struct modem_cmux *cmux, struct bg9x_pipe *pipe)
{
	struct modem_cmux_dlci dlci = &pipe->dlci;
	const struct modem_cmux_dlci_config config = {
		.dlci_address = pipe->dlci_address,
		.receive_buf = pipe->dlci_receive_buf,
		.receive_buf_size = pipe->dlci_receive_buf_size,
	};

	pipe->pipe = modem_cmux_dlci_init(cmux, dlci, &config);

	if (pipe->pipelink != NULL) {
		modem_pipelink_init(pipe->pipelink, pipe->pipe);
	}
}

static void bg9x_init_pipes(const struct device *dev)
{
	const struct bg9x_config *config = dev->config;
	struct bg9x_data *data = dev->data;
	struct modem_cmux *cmux;
	struct bg9x_pipe *pipe;

	for (size_t i = 0; i < config->dlcis_size; i++) {
		cmux = &data->cmux;
		pipe = &config->pipes[i];
		bg9x_init_pipe(cmux, pipe);
	}
}
```
Its only an snippet of the actual sample I'm working on

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	ret = modem_pipe_receive(pipe, receive_buf, sizeof(receive_buf));
+	if (ret < 1 || ret == sizeof(receive_buf)) {
```
`ret == sizeof(receive_buf)`?

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+	struct modem_pipe *pipe;
+
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
```
```suggestion
```

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+{
+	struct modem_pipe *pipe;
+
+	ARG_UNUSED(argc);
```
Check `argc`?

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+	ARG_UNUSED(argv);
+
+	if (!modem_pipelink_is_connected(shell_pipelink)) {
+		shell_print(sh, "Pipe is not connected, is modem powered on?");
```
It's gonna be messy as the modem is only powered on when the sample is doing its thing...

bjarki-trackunit replied:
The at_shell.c here will never be merged hopefully :D It is just an example of "some thing" that is listening for a pipe to become available/unavailable, and enabling/disabling itself automatically :)

bjarki-trackunit commented on include/zephyr/modem/pipelink.h:
```diff
@@ -0,0 +1,150 @@
+			   void *user_data);
+
+/**
+ * @brief Test if pipelink pipe is connected
```
Test and check can be used interchangeably for boolean operations, see `atomic_test_bit()` for example :) Changed it to Check however, I have no strong preference :)

tomi-font replied:
I think that the naming comes from the fact that the operation (bitwise and) is called test at the assembly level. Here I was thinking at a higher level. Could also be that you're right. Anyway, thanks. :slightly_smiling_face:

frkv replied:
Test is somewhat of a "weighted term". It is better to reserve that for actual testing and rely on verbs like "check", "confirm", "validate"  etc....

bjarki-trackunit replied:
Updated to use "Check whether"

bjarki-trackunit commented on include/zephyr/modem/pipelink.h:
```diff
@@ -0,0 +1,150 @@
+/** @endcond */
+
+/**
+ * @brief Attach pipelink to callback
```
Done

tomi-font commented on include/zephyr/modem/pipelink.h:
```diff
@@ -0,0 +1,150 @@
+ * @param link Pipelink instance
+ * @retval Pointer to pipe if pipelink has been initialized
+ * @retval NULL if pipelink has not been initialized
+ * @note Pipelink is guaranteed initialized after first connected event
```
```suggestion
 * @note Pipelink is guaranteed to be initialized after first connected event
```

bjarki-trackunit replied:
Note removed

tomi-font commented on samples/net/cellular_modem/src/at_shell.c:
```diff
@@ -0,0 +1,124 @@
+static uint8_t transmit_buf[255];
+static uint8_t receive_buf[255];
+
+static void receive_work_handler(struct k_work *work)
```
Claude, is that you?

bjarki-andreasen committed:
drivers: modem: add at shell

Add AT shell which uses the newly exposed pipes to
send/receive at commands from the modem identified by the
chosen node zephyr,at-shell.

To send an AT command, the shell command at is used, followed
by the command to send, and an optional response which
overwrites the default "OK"

For example, sending "AT", which returns "OK"
  modem at at <- command
  OK <- response

Enabling echo, then sending "AT" which will now return "AT" + "OK"
  modem at ati1 <- command
  OK <- response
  at at <- command
  at <- response
  OK <- response

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
samples: net: cellular_modem: include at shell

Include new AT shell alongside the PM device shell to allow
for powering back up the modem after the sample has completed
to test out the at shell.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
## Update
Rewrote the at shell properly, using an instance of modem_chat to run the command and parse the responses, and moved it to the drivers/modem folder. Also made the addition to modem_cellular way cleaner :)

bjarki-andreasen committed:
tests: subsys: modem: add test suite for pipelink

Add test suite for pipelink module

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

jukkar commented on drivers/modem/modem_at_shell.c:
```diff
@@ -0,0 +1,269 @@
+	return ret;
+}
+
+SHELL_CMD_ARG_REGISTER(at, NULL, "at <command> <response>", at_shell_cmd_handler, 1, 2);
```
As this is related to modem, should we have it under modem shell, meaning `modem at ....`?

bjarki-trackunit replied:
Updated to common `modem` shell with `at` as subcommand :) The initial reason for not doing this was to avoid conflicts with the existing `modem` shell which works with some older modem drivers, but as they are mutually exclusive, I called the new modem shell `MODEM_AT_SHELL` for now :)

jukkar reviewed (approved):
LGTM
